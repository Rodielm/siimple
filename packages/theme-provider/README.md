# @siimple/theme-provider (Experimental)

A wrapper component for providing access to your theme in all React components.

## Installation

Install this package using **npm**:

```bash
 npm install --save @siimple/theme-provider
```

## Usage

```jsx
import React from "react";
import {ThemeProvider, useTheme} from "@siimple/theme-provider";

const defaultTheme = {
    colors: {
        primary: "coral",
    },
};

const MyText = props => {
    const theme = useTheme();
    
    return (
        <span style={{color: theme.colors.primary}}>
            {props.children}
        </span>
    );
};

export default App = () => (
    <ThemeProvider theme={defaultTheme}>
        <MyText>Hello world</MyText>
    </ThemeProvider>
);

```


## API

### ThemeProvider

A simple React component that accepts a **siimple theme object** as a prop and pass it top-down using [Context](https://reactjs.org/docs/context.html).

```jsx
import {ThemeProvider} from "@siimple/theme-provider";

export default App = () => (
    <ThemeProvider theme={theme}>
        ...
    </ThemeProvider>
);
```


### useTheme

A React [Hook](https://reactjs.org/docs/hooks-intro.html) that returns the current theme provided to `<ThemeProvider>`.

```jsx
import {useTheme} from "@siimple/theme-provider";

export const Text = props => {
    const theme = useTheme();

    return (
        <span style={{color: theme.colors.primary}}>
            {props.children}
        </span>
    );
};
```


## License

[MIT License](https://github.com/jmjuanes/siimple/blob/main/LICENSE).
