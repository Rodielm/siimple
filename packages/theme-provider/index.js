import React from "react";

// Theme context
export const ThemeContext = React.createContext({});

// Use theme hook
// Returns: the current theme configuration object
export const useTheme = () => {
    return React.useContext(ThemeContext);
};

// Theme provider component
export const ThemeProvider = props => {
    let theme = useTheme();
    if (props.theme && typeof props.theme === "object" && props.theme !== theme) {
        theme = props.theme;
    }
    return React.createElement(ThemeContext.Provider, {value: theme}, props.children);
};

// HOC that wraps a component and adds the current theme as a prop
export const withTheme = Component => {
    const name = Component.displayName || Component.name || "Component";
    const ThemedComponent = React.forwardRef((props, ref) => {
        const theme = useTheme();

        return React.createElement(Component, {theme, ref, ...props}, props.children);
    });
    ThemedComponent.displayName = `withTheme(${name})`;
    return ThemedComponent;
};
