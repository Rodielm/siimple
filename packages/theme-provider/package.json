{
    "name": "@siimple/theme-provider",
    "version": "0.3.0-beta",
    "description": "A wrapper component for providing access to your theme in all React Components.",
    "author": "Josemi Juanes <josemi@siimple.xyz>",
    "type": "module",
    "license": "MIT",
    "repository": "https://github.com/jmjuanes/siimple",
    "bugs": "https://github.com/jmjuanes/siimple/issues",
    "homepage": "https://www.siimple.xyz",
    "main": "index.js",
    "keywords": [
        "siimple",
        "react",
        "theme",
        "theme-provider"
    ],
    "dependencies": {
        "react": "^17.0.2"
    }
}
