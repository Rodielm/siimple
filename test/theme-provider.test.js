/**
 * @jest-environment jsdom
 */

import renderer from "react-test-renderer";
import {ThemeProvider, useTheme} from "@siimple/theme-provider";

describe("[theme-provider] ThemeProvider component", () => {
    it("should render", () => {
        const component = renderer.create((
            <ThemeProvider theme={{}}>
                <div>Hello world</div>
            </ThemeProvider>
        ));

        expect(component.toJSON()).toMatchSnapshot();
    });
});

describe("[theme-provider] useTheme hook", () => {
    it("should return current theme", () => {
        const defaultTheme = {
            colors: {
                primary: "black",
            },
        };
        let theme = null;
        let TestComponent = () => {
            theme = useTheme();
            return null;
        };
        renderer.create((
            <ThemeProvider theme={defaultTheme}>
                <TestComponent />
            </ThemeProvider>
        ));

        expect(theme).not.toBeNull();
        expect(theme?.colors?.primary).toBe(defaultTheme.colors.primary);
    });
});
