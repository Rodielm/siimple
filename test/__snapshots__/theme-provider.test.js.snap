// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[theme-provider] ThemeProvider component should render 1`] = `
<div>
  Hello world
</div>
`;
